appinventor/components/src/com/google/appinventor/components/runtime/util/SUtil.java

```diff
@@ -0,0 +1,162 @@
+  }
+
+  /**
+   * Requests Android version specific permissions needed to connect to paired Bluetooth devices.
+   *
+   * <p>Starting with Android 12, apps must ask for BLUETOOTH_CONNECT permission rather than the
+   * deprecated BLUETOOTH and BLUETOOTH_ADMIN permissions.</p>
+   *
+   * @param form the form used for requesting permissions
+   * @param server the component requesting the permissions
+   * @param caller the calling method initiating the request, for error messages
+   * @param continuation the code to execute after the permissions have been granted
+   * @return true if a permission request is initiated or false if all required permissions are
+   *         granted
+   */
+  public static boolean requestPermissionsForAdvertising(Form form, BluetoothServer server,
+      String caller, PermissionResultHandler continuation) {
+    return requestPermissionsForS(BLUETOOTH_ADVERTISE, form, server, caller, continuation);
+  }
+
+  /**
+   * Requests Android version specific permission {@code sdk31Permission} for Bluetooth support.
+   *
+   * @param sdk31Permission the permission needed by the requesting operation
+   * @param form the form used for requesting permissions
+   * @param source the component requesting the permission
+   * @param caller the calling method initiating the request, for error messages
+   * @param continuation the code to execute after the permissions have been granted
+   * @return true if a permission request is initiated or false if all required permissions are
+   *         granted
+   */
+  public static boolean requestPermissionsForS(String sdk31Permission, Form form,
+      Component source, String caller, PermissionResultHandler continuation) {
+    return requestPermissionsForS(new String[] { sdk31Permission }, form, source, caller,
+        continuation);
+  }
+
+  /**
+   * Requests Android version specific permissions {@code sdk31Permissions} for Bluetooth support.
+   *
+   * @param sdk31Permissions the permission needed by the requesting operation
+   * @param form the form used for requesting permissions
+   * @param source the component requesting the permission
+   * @param caller the calling method initiating the request, for error messages
+   * @param continuation the code to execute after the permissions have been granted
+   * @return true if a permission request is initiated or false if all required permissions are
+   *         granted
+   */
+  public static boolean requestPermissionsForS(String[] sdk31Permissions, Form form,
+      Component source, String caller, PermissionResultHandler continuation) {
+    List<String> permsNeeded = new ArrayList<>();
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+      Collections.addAll(permsNeeded, sdk31Permissions);
+    } else {
+      permsNeeded.add(BLUETOOTH);
+      permsNeeded.add(BLUETOOTH_ADMIN);
+    }
+    return performRequest(form, source, caller, permsNeeded, continuation);
```

Comment:
As I'm working on the BLE extension I'm not 100% sure this logic is correct and I may need to come back to it.